api/s2n.h

```diff
@@ -1812,6 +1812,54 @@ S2N_API extern int s2n_connection_prefer_throughput(struct s2n_connection *conn)
  */
 S2N_API extern int s2n_connection_prefer_low_latency(struct s2n_connection *conn);
 
+/**
+ * Configure the connection to reduce potentially expensive calls to recv.
+ *
+ * If this setting is disabled, s2n-tls will call read twice for every TLS record,
+ * which can be expensive but ensures that s2n-tls will always attempt to read the
+ * exact number of bytes it requires. If this setting is enabled, s2n-tls will
+ * instead reduce the number of calls to read by attempting to read as much data
+ * as possible in each read call, storing the extra in the existing IO buffers.
+ * This may cause it to request more data than will ever actually be available.
+ *
+ * There is no additional memory cost of enabling this setting. It reuses the
+ * existing IO buffers.
+ *
+ * This setting is disabled by default. Depending on how your application detects
+ * data available for reading, buffering reads may break your event loop.
+ * In particular, note that:
+ *
+ * 1. File descriptor reads or calls to your custom s2n_recv_cb may request more
+ *    data than is available. Reads must return partial data when available rather
+ *    than blocking until all requested data is available.
+ *
+ * 2. s2n_negotiate may read and buffer application data records.
+ *    You must call s2n_recv at least once after negotiation to ensure that you
+ *    handle any buffered data.
+ *
+ * 3. s2n_recv may read and buffer more records than it parses and decrypts.
+ *    You must call s2n_recv until it reports S2N_ERR_T_BLOCKED, rather than just
+ *    until it reports S2N_SUCCESS.
+ *
+ * 4. s2n_peek reports available decrypted data. It does not report any data
+ *    buffered by this feature.
+ *
+ * 5. s2n_connection_release_buffers will not release the input buffer if it
+ *    contains buffered data.
+ *
+ * @warning This feature cannot be enabled for a connection that will enable kTLS for receiving.
+ *
+ * @warning This feature may work with blocking IO, if used carefully. Your blocking
+ * IO must support partial reads (so MSG_WAITALL cannot be used). You will need
+ * to know how much data will eventually be available rather than relying on
+ * S2N_ERR_T_BLOCKED as noted in #3 above.
+ *
```

Comment:
Should we add a note about ensuring that the recv buffer is drained by looping until `S2N_ERR_T_BLOCKED` before waiting on epoll? Otherwise the event loop could hang.

Reply:
I have a note slightly earlier in the comment: https://github.com/aws/s2n-tls/blob/29c7458cd4c44378720650c31eff548ef7edbf6a/api/s2n.h#L1840-L1842 Is that not sufficient?

This warning is more about blocking IO, which specifically can't use S2N_ERR_T_BLOCKED.

Reply:
I guess it hints at it. But I'm more saying to call out epoll directly. Like we have our own antipattern of waiting for an epoll result to tell us what to do: https://github.com/aws/s2n-tls/blob/b36c578bd1b26cf8729225d35da044e7ff941332/bin/echo.c#L373